### PR TITLE
build: switch to circleci

### PR DESCRIPTION
**Pull request by CaerusKaru**

### CODE REVIEW

**Comment by CaerusKaru** [comment]
@devversion Can you PTAL? I don't know why the ssr tests are failing. For context, they would run on Travis, but they don't run for me locally with the same error.

The publish task fails bc Jeremy has to copy over the access token.

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,37 +1,319 @@
-# Javascript Node CircleCI 2.0 configuration file
-#
-# Check https://circleci.com/docs/2.0/language-javascript/ for more details
-#
+# Configuration file for https://circleci.com/gh/angular/flex-layout
```
Comment (devversion) [inline .circleci/config.yml]: Maybe also note that this is based on Material's configuration?

File: .circleci/config.yml [hunk 2]
```diff
@@ -1,37 +1,319 @@
...
+    key: *cache_key
+
+# Saves the cache for the current Yarn lock file. We store the node modules and the build
+# cache in order to make subsequent builds faster.
+var_6: &save_cache
+  save_cache:
+    key: *cache_key
+    paths:
+      - "node_modules"
+      - "~/dist/releases"
```
Comment (devversion) [inline .circleci/config.yml]: Why would you want to cache the `release` output? The release output should be transferred through Circle's workspace storage.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,37 +1,319 @@
...
+  # ----------------------------------------------------------------------------
+  # Job that runs the unit tests on Browserstack. The browsers that will be used
+  # to run the unit tests on Browserstack are set in: test/browser-providers.js
+  # ----------------------------------------------------------------------------
+  tests_browserstack:
+    <<: *job_defaults
+    resource_class: xlarge
+    environment:
+      BROWSER_STACK_USERNAME: "angularteam1"
+      BROWSER_STACK_ACCESS_KEY: "CaXMeMHD9pr5PHg8N7Jq"
```
Comment (devversion) [inline .circleci/config.yml]: We should at some point figure out what to do about licenses.. I'm afraid that we all often run into rate limit exceptions and adding more projects might not improve that situation.

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,37 +1,319 @@
...
-      # Download and cache dependencies
-      - restore_cache:
-          keys:
-          - v1-dependencies-{{ checksum "yarn.lock" }}
-          # fallback to using the latest cache if no exact match is found
-          - v1-dependencies-
+  # -------------------------------------------------------------------------
+  # Job that makes sure Angular Layout can be integrated with a basic app
+  # -------------------------------------------------------------------------
+  hw_build:
```
Comment (devversion) [inline .circleci/config.yml]: `hw_build` is pretty unclear at first glance. Please do not abbreviate the job name.

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,37 +1,319 @@
...
+          command: |
+            mkdir -p /tmp/layout-umd-minified-bundles
+            cp dist/releases/flex-layout/bundles/*.umd.min.js /tmp/layout-umd-minified-bundles
+      # Publish bundle artifacts which will be used to calculate the size change.
+      # Note: Make sure that the size plugin from the Angular robot fetches the artifacts
+      # from this CircleCI job (see .github/angular-robot.yml). Additionally any artifacts need to
+      # be stored with the following path format: "{projectName}/{context}/{fileName}"
+      # This format is necessary because otherwise the bot is not able to pick up the
+      # artifacts from CircleCI. See:
+      # https://github.com/angular/github-robot/blob/master/functions/src/plugins/size.ts#L392-L394
```
Comment (devversion) [inline .circleci/config.yml]: Are you actually using the Github robot?

This seems to be out of scope for this PR because there is no bot configuration yet.

File: .circleci/config.yml [hunk 6]
```diff
@@ -1,37 +1,319 @@
...
+# General setup for CircleCI
+# ---------------------------
+general:
+  branches:
+    only:
+      - master
+      # 5.2.x, 6.0.x, etc
+      - /\d+\.\d+\.x/
+      # 5.x, 6.x, etc
+      - /\d+\.x/
```
Comment (devversion) [inline .circleci/config.yml]: Does `flex-layout` use publish branches? Could be probably removed.

File: .travis.yml [hunk 7]
```diff
@@ -16,49 +16,51 @@ branches:
...
-  - ./scripts/ci/travis-script.sh
-
-cache:
-  yarn: true
-  directories:
-    - ./node_modules/
+#cache:
+#  yarn: true
+#  directories:
+#    - ./node_modules/
```
Comment (devversion) [inline .travis.yml]: If this file is mostly just commented code, just remove it. Travis will be gone once that PR has been merged.

File: test/browser-providers.js [hunk 8]
```diff
@@ -2,81 +2,66 @@
 
 /*
  * Browser Configuration for the different jobs in the CI.
- * Target can be either: BS (Browserstack) | SL (Saucelabs) | TC (Travis CI) | null (To not run)
+ *
+ *   - local: Launches the browser locally on the current operating system.
+ *   - BS: Launches the browser within BrowserStack
+ *   - SL: Launches the browser within Saucelabs
+ *
+ * TODO(devversion): rename this to "browserstack" and "saucelabs".
```
Comment (devversion) [inline test/browser-providers.js]: I've already renamed this to `browserstack` and `saucelabs`. I somewhere have a PR on Material that removes this (just not merged yet).

You can remove this todo.

File: test/karma.conf.js [hunk 9]
```diff
@@ -25,18 +23,22 @@ module.exports = (config) => {
...
 
+      // Include all Angular dependencies
       {pattern: 'node_modules/@angular/**/*', included: false, watched: false},
       {pattern: 'node_modules/rxjs/**/*', included: false, watched: false},
 
       {pattern: 'test/karma-system-config.js', included: true, watched: false},
       {pattern: 'test/karma-test-shim.js', included: true, watched: false},
 
+      // Include a Material theme in the test suite.
+      {pattern: 'dist/packages/**/core/theming/prebuilt/indigo-pink.css', included: true, watched: true},
```
Comment (devversion) [inline test/karma.conf.js]: Do you actually use Material in any of your unit tests?

File: test/karma.conf.js [hunk 10]
```diff
@@ -25,18 +23,22 @@ module.exports = (config) => {
       {pattern: 'node_modules/zone.js/dist/jasmine-patch.min.js', included: true, watched: false},
       {pattern: 'node_modules/zone.js/dist/async-test.js', included: true, watched: false},
       {pattern: 'node_modules/zone.js/dist/fake-async-test.js', included: true, watched: false},
+      {pattern: 'node_modules/hammerjs/hammer.min.js', included: true, watched: false},
```
Comment (devversion) [inline test/karma.conf.js]: Same as for Material, do you use hammer as well as moment in unit tests?

File: test/protractor.conf.js [hunk 11]
```diff
@@ -24,37 +24,24 @@ const config = {
       path: '../tools/axe-protractor/axe-protractor.js',
 
       rules: [
-        // Exclude md-menu elements because those are empty if not active.
-        { id: 'aria-required-children', selector: '*:not(md-menu)' },
+        // Exclude mat-menu elements because those are empty if not active.
+        { id: 'aria-required-children', selector: '*:not(mat-menu)' },
```
Comment (devversion) [inline test/protractor.conf.js]: You should just remove these rules for flex-layout. There is no test for `mat-menu` AFAIK.

**Comment by devversion** [comment]
@CaerusKaru Before this can be merged, the `tests_ssr` job should definitely be green, right?

cc. @jelbourn for the CircleCI publish token.

File: .circleci/config.yml (hunk 1 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: Yes definitely. I'll also credit you explicitly because this work is fantastic.

File: .circleci/config.yml (hunk 3 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: We have our own, and I'll add it here to avoid using up the main one, but I just wanted to validate the format first

File: .circleci/config.yml (hunk 4 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: Will change to `hello_world_build`

File: .circleci/config.yml (hunk 5 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: I think we'd like to start, if anything to track size regressions. I figure we could add the config here and then onboard with the robot later?

File: .travis.yml (hunk 7 above)
Comment (CaerusKaru) [reply .travis.yml]: Travis is still required, so if this gets merged before that is turned off, it still needs to run. If Jeremy turns off Travis pre-merge I'll delete the file, ow will remove in a follow-up

File: test/karma.conf.js (hunk 9 above)
Comment (CaerusKaru) [reply test/karma.conf.js]: Yes we do! I know it's surprising, though we probably don't need this theme.

File: .circleci/config.yml (hunk 3 above)
Comment (devversion) [reply .circleci/config.yml]: Cool, good to know!

File: test/karma.conf.js (hunk 10 above)
Comment (CaerusKaru) [reply test/karma.conf.js]: Most likely not, will remove

File: test/protractor.conf.js (hunk 11 above)
Comment (CaerusKaru) [reply test/protractor.conf.js]: Yeah I think our only Material test dependency is on button. Either way, will remove

File: .circleci/config.yml (hunk 2 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: Will follow-up on this one offline

**Comment by CaerusKaru** [comment]
@devversion Thanks for the review, SSR tests definitely need to be green for this to get merged. I'll work on it.

File: .circleci/config.yml (hunk 5 above)
Comment (devversion) [reply .circleci/config.yml]: Yeah that works.

**Comment by CaerusKaru** [comment]
@devversion resolved all comments, PTAL :+1:

**Comment by CaerusKaru** [comment]
Also marking as blocked until SSR issues can be resolved

File: .circleci/config.yml [hunk 12]
```diff
@@ -1,37 +1,320 @@
...
+  # ----------------------------------------------------------------------------
+  # Job that runs the unit tests on Saucelabs. The browsers that will be used
+  # to run the unit tests on Saucelabs are set in: test/browser-providers.js
+  # ----------------------------------------------------------------------------
+  tests_saucelabs:
+    <<: *job_defaults
+    resource_class: xlarge
+    environment:
+      SAUCE_USERNAME: "angular-ci"
+      SAUCE_ACCESS_KEY: "9b988f434ff8-fbca-8aa4-4ae3-35442987"
```
Comment (devversion) [inline .circleci/config.yml]: What about that license? Nice in regards to BrowserStack

File: .circleci/config.yml [hunk 13]
```diff
@@ -1,37 +1,320 @@
...
+
+      - run: ./scripts/circleci/run-browserstack-tests.sh
+
+  # ----------------------------------------------------------------------------
+  # Job that runs the unit tests on Saucelabs. The browsers that will be used
+  # to run the unit tests on Saucelabs are set in: test/browser-providers.js
+  # ----------------------------------------------------------------------------
+  tests_saucelabs:
+    <<: *job_defaults
+    resource_class: xlarge
```
Comment (devversion) [inline .circleci/config.yml]: Flex-layout should probably go down with the resource class here. On Material we have it enabled for tests because we have around 3k. `Medium` is the default and should be good enough, thoughts?

(same applies for browserstack, local browsers and saucelabs)

File: .circleci/config.yml (hunk 12 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: Discussed offline, will not switch for SauceLabs at the moment. Could revisit in the future

File: .circleci/config.yml (hunk 13 above)
Comment (CaerusKaru) [reply .circleci/config.yml]: I kind of like that it runs super-quick though 😉 . But you're probably right.  The default should probably be plenty. I'll try it out in a fixup commit, and if it's not good enough I'll revert back
